### PR TITLE
Add unsupported kind values of MethodHandle

### PR DESCRIPTION
**Pull request by agorski3**
See commits for details

### CODE REVIEW

**Comment by JesusFreke** [comment]
Interesting. Looks like the docs at https://source.android.com/devices/tech/dalvik/dex-format#method-handle-type-codes need to be updated :)

**Comment by agorski3** [comment]
Yep. From what I can tell those are the ones supported by the now defunct Jack compiler but everywhere in android 8 that I have seen them includes all 9.

**Comment by agorski3** [comment]
The import path was wrong since the refactoring. Unrelated but figured you would want to fix it.

File: dexlib2/src/main/java/org/jf/dexlib2/MethodHandleType.java [hunk 1]
```diff
@@ -31,59 +31,62 @@
...
 
 public class MethodHandleType {
-    public static final int STATIC_PUT = 0;
-    public static final int STATIC_GET = 1;
-    public static final int INSTANCE_PUT = 2;
-    public static final int INSTANCE_GET = 3;
-    public static final int STATIC_INVOKE = 4;
-    public static final int INSTANCE_INVOKE = 5;
+    /** A method handle that gets an instance field. */
+    public static final int REF_GET_FIELD = 3;
```
Comment (JesusFreke) [inline dexlib2/src/main/java/org/jf/dexlib2/MethodHandleType.java]: Let's stick with the original naming scheme, which more closely matches the canonical dex specification at https://source.android.com/devices/tech/dalvik/dex-format. Unless you're planning on updating the constants there as well?

Although now that I'm looking at it, maybe switch STATIC_INVOKE, etc. to INVOKE_STATIC, to better match the constants there. Up to you though, I'm happy to make that change afterward if you prefer.

File: dexlib2/src/main/java/org/jf/dexlib2/MethodHandleType.java [hunk 2]
```diff
@@ -31,59 +31,62 @@
...
-        Integer ret = methodHandleTypeNames.get(methodHandleType);
+        Integer ret = methodHandleTypeNames.inverse().get(methodHandleType);
         if (ret == null) {
             throw new ExceptionWithContext("Invalid method handle type: %s", methodHandleType);
         }
         return ret;
     }
 
     public static class InvalidMethodHandleTypeException extends ExceptionWithContext {
+        private static final long serialVersionUID = -2019141391365188343L;
```
Comment (JesusFreke) [inline dexlib2/src/main/java/org/jf/dexlib2/MethodHandleType.java]: Nothing else sets this, so let's leave it out to be consistent. If it becomes a problem for some reason, I would prefer to add these to all of the custom exception/serializable classes in one fell swoop.

File: dexlib2/src/main/java/org/jf/dexlib2/MethodHandleType.java (hunk 1 above)
Comment (agorski3) [reply dexlib2/src/main/java/org/jf/dexlib2/MethodHandleType.java]: I disagree. The naming scheme in dex is poor and confusing. Why they decided to change it from what is used in normal java I do not understand. These are much clearer. Plus with the comment above the field, there should be no confusion on what they are pointing to.

File: dexlib2/src/main/java/org/jf/dexlib2/MethodHandleType.java (hunk 2 above)
Comment (agorski3) [reply dexlib2/src/main/java/org/jf/dexlib2/MethodHandleType.java]: This I can change. I was probably just trying to get rid of warnings and forgot I did it.

File: dexlib2/src/main/java/org/jf/dexlib2/MethodHandleType.java (hunk 1 above)
Comment (JesusFreke) [reply dexlib2/src/main/java/org/jf/dexlib2/MethodHandleType.java]: dexlib2 is a parser for the dex format, not for java bytecode :). the dex specification is the canonical source for the dex format, and I'd prefer to stay consistent with that.

Comment (JesusFreke) [reply dexlib2/src/main/java/org/jf/dexlib2/MethodHandleType.java]: IMO, the dex constants make more sense in the context of the dex bytecode.

e.g. there is no other reference to the "special" method nomenclature (invokespecial, etc.) in the context of a dex file. The equivalent appears to be invoke-direct. Why would we use ref_invoke_special, when the equivalent bytecode is invoke-direct?

Comment (JesusFreke) [reply dexlib2/src/main/java/org/jf/dexlib2/MethodHandleType.java]: Also STATIC_GET = sget, INSTANCE_GET = iget, etc.

while the "GET_FIELD" and "PUT_FIELD" nomenclature appears to be referencing java's getfield and putfield instructions.

Comment (agorski3) [reply dexlib2/src/main/java/org/jf/dexlib2/MethodHandleType.java]: Alright I guess. I still find the names unclear but I can put them in the exact form that dexdump uses (including the strings) which matches the ones on the dex-format page minus those missing. This is what I mean here

https://android.googlesource.com/platform/art/+/android-8.1.0_r41/dexdump/dexdump.cc#1583

Comment (JesusFreke) [reply dexlib2/src/main/java/org/jf/dexlib2/MethodHandleType.java]: Yeah, let's keep the existing ones as-is (unless you want to change `*_INVOKE` to `INVOKE_*` -- up to you), and then the new ones should be INVOKE_DIRECT, INVOKE_CONSTRUCTOR and INVOKE_INTERFACE.

**Comment by agorski3** [comment]
Ok it should be updated now how you wanted. To get the smali parser to work correctly I added mht- (i.e. method handle type) in front of all the string representations of the values because otherwise it failed on things like invoke-static (since that is a valid instruction).

**Comment by agorski3** [comment]
Oh and I did not notice the spacing issues (tabs instead of spaces) until after pushing. It is fixed now too. I just squashed it all into the last commit.

**Comment by JesusFreke** [comment]
Thanks! I think we should be able to keep the "invoke-static", etc. tokens for this without ambiguity, since they are used in different contexts. I wonder if that's why I decided to go with "static-invoke" instead of "invoke-static" at the time. I honestly can't remember.. and it hasn't even been that long ago. hah!

In any case, I do like the "invoke-static" style token for consistency with the constants in the dex-format, and also with their corresponding instructions.

See, e.g. how the 12x format is split out into INSTRUCTION_FORMAT12x and INSTRUCTION_FORMAT12x_OR_ID, for the instruction names that also happen to be valid identifiers.

Similarly, we could do something like:

    "invoke-super" {
      return newToken(INSTRUCTION_FORMAT35c_METHOD);
    }

    "invoke-virtual" | "invoke-direct" | "invoke-static" | "invoke-interface" {
        return newToken(INSTRUCTION_FORMAT35c_METHOD_OR_METHOD_HANDLE_TYPE);
    }



And on a related note, I have a pending change out to update the list of method handle types on https://source.android.com/devices/tech/dalvik/dex-format.

**Comment by agorski3** [comment]
Ok that should fix the parser so it properly parses tokens for method handle types and invoke instructions which have the same value.

**Comment by agorski3** [comment]
Oops. Thanks Fixed that.

**Comment by JesusFreke** [comment]
Sorry for the delay :) I was travelling and without internet access for a bit last/this week. I'll try and take a hopefully final look at this today or this weekend and get it merged.

And in other news, the dex-format document has been updated to include the missing method handle types: https://source.android.com/devices/tech/dalvik/dex-format#method-handle-type-codes

**Comment by JesusFreke** [comment]
This was merged as 86963ccb43b7be83d5d333854e14af51e8e84ba7, after a bit of rebasing and minor tweaks.

**Pull request closed without merging**